oauth2-core/src/main/java/org/springframework/security/oauth2/core/OAuth2Error.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2012-2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.core;
+
+import org.springframework.http.HttpStatus;
+
+import java.util.Arrays;
+import java.util.Optional;
+
+/**
+ * @author Joe Grandja
+ */
+public final class OAuth2Error {
+	private final ErrorCode errorCode;
+	private final String description;
+	private final String uri;
+	private final HttpStatus statusCode;
+
+	public enum ErrorCode {
```

Comment:
I struggle with this being an Enum because I don't think we will ever be able to guarantee all the errors are enumerated (which is what an Enum is). We could potentially allow OAuth2Error to accept an ErrorCode interface and then create some constants for well known errors